test(subscription) : Subscription flow tests using playwright

ashrivastava-qa opened the pull request:
## Because
- added test for new subscription using credit card
- added test for new subscription using test paypal account
- added test for new subscription after failed subscription to address  #11731 

## This pull request

- Contains the first 4 high priority test as discussed with the team in [Subplat automation plan draft](https://docs.google.com/document/d/1El1rsBG7LZ3xX3T-N6FF0Yz5SDGEqFUfq21gH6jFe8I/edit#) doc

## Issue that this pull request solves

Closes:  #11731 

## Checklist

_Put an `x` in the boxes that apply_

- [X] My commit is GPG signed.
- [X] If applicable, I have modified or added tests which pass locally.
- [X] I have added necessary documentation (if appropriate).
- [X] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

@mozilla/fxa-devs please review! thanks!

pdehaan commented on packages/functional-tests/pages/products/index.ts:
```diff
@@ -7,16 +9,50 @@ export class SubscribePage extends BaseLayout {
 
   async setCreditCardInfo() {
     const frame = this.page.frame({ url: /elements-inner-card/ });
+    await frame.fill('.InputElement[name=cardnumber]', '');
```
Q: Why set an empty cardnumber before filling it again below on L13?

ashrivastava-qa replied:
This is because I used this same function for the third test "subscribe with CC after initial failed subscription" and after entering an incorrect cc number the test was failing on the step to replace the CC number with correct details. So I had to delete the previous CC details(only cc number) and then re-enter it.

pdehaan commented on packages/functional-tests/pages/products/index.ts:
```diff
@@ -7,16 +9,50 @@ export class SubscribePage extends BaseLayout {
+    );
+    await paypalWindow.click('button[id=btnNext]');
+    await paypalWindow.waitForLoadState('load');
+    await paypalWindow.fill('input[type=password]', 'Ah4SboP6UDZx95I');
```
Dito re: env var, plus now we need to delete this account and get new creds.

ashrivastava-qa replied:
Why? We can still reuse this, right?

pdehaan commented on packages/functional-tests/pages/products/index.ts:
```diff
@@ -7,16 +9,50 @@ export class SubscribePage extends BaseLayout {
+  }
+
+  async clickTryAgain() {
+    return Promise.all([this.page.click('[data-testid="retry-link"]')]);
```
nit: not sure if this needs a `Promise.all()` wrapper if it's the only promise in an array.

Maybe just this?
```js
return this.page.click('[data-testid="retry-link"]');
```

pdehaan commented on packages/functional-tests/tests/misc.spec.ts:
```diff
@@ -1,69 +1,48 @@
 import { test, expect } from '../lib/fixtures/standard';
 
-test.describe('severity-1', () => {
```
I'm slightly confused here... We have a `test.describe("severity-1", ...)` wrapper here which will let us glob by "severity-1" keyword, presumably. THis PR seems to remove that and bumps the tests up to top level. 
Do we not need to filter by "severity-1" anymore, or am I misreading something?

ashrivastava-qa replied:
Yes you are right. I am going to add the wrapper again here.

pdehaan commented on packages/functional-tests/tests/subscription.spec.ts:
```diff
@@ -0,0 +1,59 @@
+  });
+});
+
+test.describe('severity-1', () => {
```
We have this as well on L3 above.

Should we just do something like:
```js
test.describe("severity-1", () => {
  test("...", asyc () => {...});
  test("...", asyc () => {...});
  test("...", asyc () => {...});
});
```

pdehaan commented on packages/functional-tests/tests/subscription.spec.ts:
```diff
@@ -0,0 +1,59 @@
+import { test, expect } from '../lib/fixtures/standard';
+
+test.describe('severity-1', () => {
+  test('subscribe with credit card and login to product', async ({
```
nit: we say "credit card" here. should we clarify to "stripe credit card"?
Not sure it matters, or if we ever want to grep by "stripe" to run specific tests.

ashrivastava-qa replied:
good suggestion, not sure if it's required but good to have. I will add it.

pdehaan commented on packages/functional-tests/tests/subscription.spec.ts:
```diff
@@ -0,0 +1,59 @@
+});
+
+test.describe('severity-1', () => {
+  test('subscribe with paypal and login to product', async ({
```
nit: I might suggest moving this to the first test in the `severity-1` suite.
That would give us the following (where the 2 stripe-specific tests are grouped a bit closer):
1. subscribe with paypal and login to product
2. subscribe with [stripe] credit card and login to product
3. subscribe with [stripe] credit card after initial failed subscription

pdehaan commented on packages/functional-tests/tests/subscription.spec.ts:
```diff
@@ -0,0 +1,59 @@
+  test('subscribe with credit card and login to product', async ({
+    pages: { relier, login, subscribe },
+  }, { project }) => {
+    test.skip(project.name === 'production', 'prod needs a valid credit card');
+    test.skip(project.name === 'local', 'No need to be run on local');
```
nit: Wish there was an easier way to simplify this instead of having to duplicate it for each test, but not sure that the `test.describe` has access to the project names/envs.
Even if we refactored out into a helper function, not sure that helps w/ DRYness or clarity.

ashrivastava-qa replied:
I know, I tried a few things but nothing worked. So, for now I am keeping it as is but will come back later if I find something and rewrite it.

pdehaan commented on packages/functional-tests/pages/products/index.ts:
```diff
@@ -1,4 +1,6 @@
 import { BaseLayout } from '../layout';
+import { BrowserContext } from 'playwright';
+let context: BrowserContext;
```
Unclear on where/if this is used. I see these 2 instances of "BrowserContext" and all 3 instances of "context" seem to be on these 2 lines. I didn't see any other instances of `context.`, so not sure if we're using this anywhere.

ashrivastava-qa replied:
Yeah. Good catch! I added that to be used for the earlier issue with Paypal frame not being accessible, to try and see if context would help but forgot to remove it since I didn't use it. I will remove it. Thanks!

pdehaan reviewed (commented):
Mostly nits (which you can take or ignore).

pdehaan commented:
https://www.npmjs.com/package/playwright seems to say latest is [1.20.1](https://github.com/microsoft/playwright/releases/tag/v1.20.1) (which might have been published 8 days ago).
[1.20.0](https://github.com/microsoft/playwright/releases/tag/v1.20.0) was published 17 days ago.

<https://github.com/microsoft/playwright/issues/12081> might be promising (also <https://github.com/microsoft/playwright/issues/12367>, but i think they are the same core issue of too-many-playwrights).

ashrivastava-qa commented:
CircleCi failure solution:
- Updated the playwright and @playwright/test version in the package.json file to 1.20.0
- Updated the playwright docker image version in config.yaml file

vbudhram commented on packages/functional-tests/tests/subscription.spec.ts:
```diff
@@ -0,0 +1,55 @@
+    expect(await relier.isPro()).toBe(true);
+  });
+
+  test('subscribe with credit card after initial failed subscription', async ({
```
I noticed that some tests reference a corresponding test rail test case. Do these have any?

ashrivastava-qa commented:
@vbudhram Thanks for the review! Yes that stage failure is because of a bug that is currently filed by QA: [https://mozilla-hub.atlassian.net/browse/FXA-4816].

vbudhram reviewed (approved):
@ashrivastava-qa Thanks!
